src/sentry/debug_files/artifact_bundles.py

```diff
@@ -34,14 +50,374 @@
+
+        try:
+            index._build()
+            return index
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_building")
+            return None
+
+    def _build(self) -> None:
+        loaded_index = self.store.load()
+
+        # If we don't find an index, we will silently not mutate the state, since it's initialized as empty.
+        if loaded_index is not None:
+            self._bundles = cast(Bundles, loaded_index.get("bundles", []))
+            self._files_by_url = cast(FilesByUrl, loaded_index.get("files_by_url", {}))
+            self._files_by_debug_id = cast(
+                FilesByDebugID, loaded_index.get("files_by_debug_id", {})
+            )
+
+    def merge(self, artifact_bundle: ArtifactBundle, archive: ArtifactBundleArchive) -> bool:
+        try:
+            # We want to index by url if the store has an identifier which is being indexed by release.
+            self._merge(artifact_bundle, archive, self.store.identifier.is_indexing_by_release())
+            return True
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_merging")
+            return False
+
+    def _merge(
+        self, artifact_bundle: ArtifactBundle, archive: ArtifactBundleArchive, index_urls: bool
+    ):
+        bundle_index = self._update_bundles(artifact_bundle)
+
+        if index_urls:
+            self._iterate_over_urls(bundle_index, archive)
+        else:
+            self._iterate_over_debug_ids(bundle_index, archive)
+
+    def _update_bundles(self, artifact_bundle: ArtifactBundle) -> int:
+        if artifact_bundle.date_last_modified is None:
+            raise Exception("Can't index a bundle with no date last modified")
+
+        bundle_meta = BundleMeta(
+            id=artifact_bundle.id, timestamp=artifact_bundle.date_last_modified
+        )
+
+        bundle_index = self._index_of_bundle(artifact_bundle.id)
+        if bundle_index is None:
+            self._bundles.append(bundle_meta)
+            return len(self._bundles) - 1
+        else:
+            self._bundles[bundle_index] = bundle_meta
+            return bundle_index
+
+    def _iterate_over_debug_ids(self, bundle_index: int, archive: ArtifactBundleArchive):
+        for debug_id, _ in archive.get_all_debug_ids():
+            self._add_sorted_entry(self._files_by_debug_id, debug_id, bundle_index)
+
+    def _iterate_over_urls(self, bundle_index: int, archive: ArtifactBundleArchive):
+        for url in archive.get_all_urls():
+            self._add_sorted_entry(self._files_by_url, url, bundle_index)
+
+    def _add_sorted_entry(self, collection: Dict[T, List[int]], key: T, bundle_index: int):
+        entries = collection.get(key, [])
+        entries.append(bundle_index)
+        entries.sort(key=lambda index: self._bundles[index]["timestamp"], reverse=True)
```

Comment:
The consumer side (aka symbolicator) is expecting the *last* item to be the most recent one, so a straight sort (not reversed) should be used here.